<a href="https://colab.research.google.com/github/heatherale/data_course/blob/main/hw_3_2_%D0%9E%D1%81%D0%BD%D0%BE%D0%B2%D0%B8_%D1%80%D0%BE%D0%B1%D0%BE%D1%82%D0%B8_%D0%B7_%D0%B4%D0%B0%D0%BD%D0%B8%D0%BC%D0%B8_%D1%83_Pandas_Khudolii_Kateryna.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd
data_path = '/content/drive/MyDrive/Pandas homeworks/data/supermarket_sales.csv'
df = pd.read_csv(data_path)
df

Invoice ID Branch       City Customer type  Gender  \
0    750-67-8428      A     Yangon        Member  Female   
1    226-31-3081      C  Naypyitaw        Normal  Female   
2    631-41-3108      A     Yangon        Normal    Male   
3    123-19-1176      A     Yangon        Member    Male   
4    373-73-7910      A     Yangon        Normal    Male   
..           ...    ...        ...           ...     ...   
995  233-67-5758      C  Naypyitaw        Normal    Male   
996  303-96-2227      B   Mandalay        Normal  Female   
997  727-02-1313      A     Yangon        Member    Male   
998  347-56-2442      A     Yangon        Normal    Male   
999  849-09-3807      A     Yangon        Member  Female   

               Product line  Unit price  Quantity   Tax 5%      Total  \
0         Health and beauty       74.69         7  26.1415   548.9715   
1    Electronic accessories       15.28         5   3.8200    80.2200   
2        Home and lifestyle       46.33         7  16.2155   340.5255   
3         Health and beauty       58.22         8  23.2880   489.0480   
4         Sports and travel       86.31         7  30.2085   634.3785   
..                      ...         ...       ...      ...        ...   
995       Health and beauty       40.35         1   2.0175    42.3675   
996      Home and lifestyle       97.38        10  48.6900  1022.4900   
997      Food and beverages       31.84         1   1.5920    33.4320   
998      Home and lifestyle       65.82         1   3.2910    69.1110   
999     Fashion accessories       88.34         7  30.9190   649.2990   

          Date   Time      Payment    cogs  gross margin percentage  \
0     1/5/2019  13:08      Ewallet  522.83                 4.761905   
1     3/8/2019  10:29         Cash   76.40                 4.761905   
2     3/3/2019  13:23  Credit card  324.31                 4.761905   
3    1/27/2019  20:33      Ewallet  465.76                 4.761905   
4     2/8/2019  10:37      Ewallet  604.17                 4.761905   
..         ...    ...          ...     ...                      ...   
995  1/29/2019  13:46      Ewallet   40.35                 4.761905   
996   3/2/2019  17:16      Ewallet  973.80                 4.761905   
997   2/9/2019  13:22         Cash   31.84                 4.761905   
998  2/22/2019  15:33         Cash   65.82                 4.761905   
999  2/18/2019  13:28         Cash  618.38                 4.761905   

     gross income  Rating  
0         26.1415     9.1  
1          3.8200     9.6  
2         16.2155     7.4  
3         23.2880     8.4  
4         30.2085     5.3  
..            ...     ...  
995        2.0175     6.2  
996       48.6900     4.4  
997        1.5920     7.7  
998        3.2910     4.1  
999       30.9190     6.6  

[1000 rows x 17 columns]

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [3]:
df.head(4)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4

In [4]:
df.tail(4)

Invoice ID Branch      City Customer type  Gender         Product line  \
996  303-96-2227      B  Mandalay        Normal  Female   Home and lifestyle   
997  727-02-1313      A    Yangon        Member    Male   Food and beverages   
998  347-56-2442      A    Yangon        Normal    Male   Home and lifestyle   
999  849-09-3807      A    Yangon        Member  Female  Fashion accessories   

     Unit price  Quantity  Tax 5%     Total       Date   Time  Payment  \
996       97.38        10  48.690  1022.490   3/2/2019  17:16  Ewallet   
997       31.84         1   1.592    33.432   2/9/2019  13:22     Cash   
998       65.82         1   3.291    69.111  2/22/2019  15:33     Cash   
999       88.34         7  30.919   649.299  2/18/2019  13:28     Cash   

       cogs  gross margin percentage  gross income  Rating  
996  973.80                 4.761905        48.690     4.4  
997   31.84                 4.761905         1.592     7.7  
998   65.82                 4.761905         3.291     4.1  
999  618.38                 4.761905        30.919     6.6

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [6]:
df[615:620:1]

Invoice ID Branch       City Customer type  Gender  \
615  509-29-3912      A     Yangon        Member  Female   
616  327-40-9673      B   Mandalay        Member    Male   
617  840-19-2096      C  Naypyitaw        Member    Male   
618  828-46-6863      A     Yangon        Member    Male   
619  641-96-3695      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
615       Sports and travel       38.72         9  17.4240  365.9040   
616       Sports and travel       72.60         6  21.7800  457.3800   
617  Electronic accessories       87.91         5  21.9775  461.5275   
618      Food and beverages       98.53         6  29.5590  620.7390   
619     Fashion accessories       43.46         6  13.0380  273.7980   

          Date   Time      Payment    cogs  gross margin percentage  \
615  3/20/2019  12:24      Ewallet  348.48                 4.761905   
616  1/13/2019  19:51         Cash  435.60                 4.761905   
617  3/14/2019  18:10      Ewallet  439.55                 4.761905   
618  1/23/2019  11:22  Credit card  591.18                 4.761905   
619   2/7/2019  17:55      Ewallet  260.76                 4.761905   

     gross income  Rating  
615       17.4240     4.2  
616       21.7800     6.9  
617       21.9775     4.4  
618       29.5590     4.0  
619       13.0380     8.5

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1Otufe_TXRmjZxWigzBiVBLCAVUfEReXM)

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [7]:
df.dtypes

Invoice ID                  object
Branch                      object
City                        object
Customer type               object
Gender                      object
Product line                object
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                        object
Time                        object
Payment                     object
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object

Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [8]:
df['gross income'].describe()

count    1000.000000
mean       15.379369
std        11.708825
min         0.508500
25%         5.924875
50%        12.088000
75%        22.445250
max        49.650000
Name: gross income, dtype: float64

Значення 75-го персентилю колонки Gross Income - 22.44525. Це означає, що 75% даних у цій колонці менші за це значення.

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1bDa1YFUWDsPm78Lnpy2geQND8M1HEQiT)

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.



  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [9]:
df['City'].value_counts()

City
Yangon       340
Mandalay     332
Naypyitaw    328
Name: count, dtype: int64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1G8c6M875NOW7oU3OtZegTOnyD2y8lhAV)

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [11]:
df['City'].value_counts(normalize='True')

City
Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: proportion, dtype: float64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1yZZpWhL2y8DX9rSRioYCRh1VbJgcf3fY)

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [15]:
print("Average gross income in Yangon:")
print(df[df.City=="Yangon"]["gross income"].mean())
print("Average gross income in Mandalay:")
print(df[df.City=="Mandalay"]["gross income"].mean())

Average gross income in Yangon:
14.874001470588235
Average gross income in Mandalay:
15.232024096385542


Чому середній валовий дохід в місті Mandalay вищий за середній валовий дохід в місті Yangon?

Гіпотеза 1: в місті Yangon продається менше товарів більш "прибуткових" категорій, ніж в місті Mandalay, або середній чек в цих категоріях менший. Перевірити це можна, перевіривши середній валовий дохід за категоріями:

In [17]:
df.groupby(['Product line','City'])['gross income'].mean()

Product line            City     
Electronic accessories  Mandalay     14.763155
                        Naypyitaw    16.423355
                        Yangon       14.537392
Fashion accessories     Mandalay     12.606234
                        Naypyitaw    15.794923
                        Yangon       15.249775
Food and beverages      Mandalay     14.490370
                        Naypyitaw    17.147803
                        Yangon       14.091216
Health and beauty       Mandalay     17.952075
                        Naypyitaw    15.215500
                        Yangon       12.763681
Home and lifestyle      Mandalay     16.713490
                        Naypyitaw    14.704289
                        Yangon       16.422854
Sports and travel       Mandalay     15.351919
                        Naypyitaw    16.679289
                        Yangon       15.635754
Name: gross income, dtype: float64

Дійсно бачимо, що в Mandalay на порядок вищий середній валовий дохід в категорії Health and beauty. Розглянемо кількість інвойсів в категоріях:

In [20]:
df.groupby(['Product line','City'])['Invoice ID'].count()

Product line            City     
Electronic accessories  Mandalay     55
                        Naypyitaw    55
                        Yangon       60
Fashion accessories     Mandalay     62
                        Naypyitaw    65
                        Yangon       51
Food and beverages      Mandalay     50
                        Naypyitaw    66
                        Yangon       58
Health and beauty       Mandalay     53
                        Naypyitaw    52
                        Yangon       47
Home and lifestyle      Mandalay     50
                        Naypyitaw    45
                        Yangon       65
Sports and travel       Mandalay     62
                        Naypyitaw    45
                        Yangon       59
Name: Invoice ID, dtype: int64

Як бачимо, річ дійсно в розмірі середнього чека в певній категорії - кількість продажів по категоріях приблизно однакова. Що ж може впливати на середній чек? Гіпотеза #2 - в місті Mandalay більше покупок здійснюють members, аніж normal customers.

In [21]:
df.groupby(['Product line','City', 'Customer type'])['Invoice ID'].count()

Product line            City       Customer type
Electronic accessories  Mandalay   Member           27
                                   Normal           28
                        Naypyitaw  Member           20
                                   Normal           35
                        Yangon     Member           31
                                   Normal           29
Fashion accessories     Mandalay   Member           32
                                   Normal           30
                        Naypyitaw  Member           32
                                   Normal           33
                        Yangon     Member           22
                                   Normal           29
Food and beverages      Mandalay   Member           29
                                   Normal           21
                        Naypyitaw  Member           36
                                   Normal           30
                        Yangon     Member           29
                                   Normal           29
Health and beauty       Mandalay   Member           26
                                   Normal           27
                        Naypyitaw  Member           25
                                   Normal           27
                        Yangon     Member           22
                                   Normal           25
Home and lifestyle      Mandalay   Member           21
                                   Normal           29
                        Naypyitaw  Member           28
                                   Normal           17
                        Yangon     Member           34
                                   Normal           31
Sports and travel       Mandalay   Member           30
                                   Normal           32
                        Naypyitaw  Member           28
                                   Normal           17
                        Yangon     Member           29
                                   Normal           30
Name: Invoice ID, dtype: int64

Ця гіпотеза теж не підтвердилась - кількість покупок між member та normal приблизно однакова.
Гіпотеза #3 - в місті Mandalay середня ціна товарів, що продаються, більша.

In [24]:
df.groupby(['Product line','City'])['Unit price'].mean()

Product line            City     
Electronic accessories  Mandalay     49.854182
                        Naypyitaw    55.809455
                        Yangon       54.871167
Fashion accessories     Mandalay     54.843871
                        Naypyitaw    59.736000
                        Yangon       56.670392
Food and beverages      Mandalay     55.540000
                        Naypyitaw    57.273030
                        Yangon       54.974483
Health and beauty       Mandalay     58.185660
                        Naypyitaw    55.971346
                        Yangon       49.862340
Home and lifestyle      Mandalay     55.514000
                        Naypyitaw    54.334222
                        Yangon       55.845692
Sports and travel       Mandalay     59.678065
                        Naypyitaw    55.107333
                        Yangon       55.610339
Name: Unit price, dtype: float64

І дійсно це так. В найбільш "дорогій" категорії в місті Mandalay товари в середньому дорожчі на 9 валютних одиниць. Чи це повʼязано з більш складною логістикою, чи більш преміальним сегментом товарів, представленим в магазині, можна дізнатись із додаткових даних.

Одна із гіпотез підтвердилась, в подальшому я б ще перевірила розмір податку (є країни, де податок залежить від області або штату), рейтинг (можливо більш задоволені покупці частіше повертаються і більше купують, і можливо в Mandalay краща якість обслуговування), середній валовий дохід по різним типам покупців (можливо, в Mandalay частіше купують жінки, тому товарів з beauty категорії продається більше).

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1JdXUmK46WcoXUJYP7NS2SJlr2XjBbpQ8)

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [25]:
df['gross margin percentage'].value_counts()

gross margin percentage
4.761905    1000
Name: count, dtype: int64

маржа однакова в усіх інвойсах.

Очікуваний результат: відповідь щодо маржі в усіх інвойсах на основі отриманих результатів

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [37]:
print("Sales by category in Yangon")
df.query('City=="Yangon"')['Product line'].value_counts(normalize=True)

Sales by category in Yangon


Product line
Home and lifestyle        0.191176
Electronic accessories    0.176471
Sports and travel         0.173529
Food and beverages        0.170588
Fashion accessories       0.150000
Health and beauty         0.138235
Name: proportion, dtype: float64

In [38]:
print("Sales by category in Mandalay")
df.query('City=="Mandalay"')['Product line'].value_counts(normalize=True)

Sales by category in Mandalay


Product line
Fashion accessories       0.186747
Sports and travel         0.186747
Electronic accessories    0.165663
Health and beauty         0.159639
Food and beverages        0.150602
Home and lifestyle        0.150602
Name: proportion, dtype: float64

Очікуваний результат для кожного з міст:

![](https://drive.google.com/uc?export=view&id=1IbG58YK5KVqC-wCnllF1btgEgfWTbi_O)

![](https://drive.google.com/uc?export=view&id=1aKOkLmlMq1sOh4GbwyaQyjfhNJzdU4Yh)